analyze.py

```diff
@@ -388,8 +388,24 @@ def dump_json_file(dir, data, json_file_name):
     json.dump(data, open(os.path.join(
         dir, json_file_name), "w"), indent=4, sort_keys=True)
 
-
-def analyze_trace(project, dir, mode, generate_oracle=True, generate_config=True, two_sided=False, use_sql=True):
+def side_effect_filter(causality_pairs, event_key_map):
+    filtered_causality_pairs = []
+    for pair in causality_pairs:
+        start_event = pair[0]
+        side_effect = pair[1] # e.g. delete sth
+        flag = False
+        for event in event_key_map[side_effect.key]:
+            # We will find add sth
+            if event.start_timestamp <= side_effect.end_timestamp:
+                continue
+            if event.etype == "Added":
```

Comment:
You might also want to check whether the side effect is a delete or add (tho currently we are only considering delete)